### PR TITLE
Make connection secret propagation UID-agnostic

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Fixes https://github.com/crossplane/crossplane-runtime/issues/140

This allows propagation to function even when the propagating and/or propagated secrets have been deleted and recreated, and thus allocated new UIDs.

### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the RBAC permissions in [`clusterrole.yaml`] to include any new types.

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplane/crossplane/tree/master/PendingReleaseNotes.md
[`clusterrole.yaml`]: https://github.com/crossplane/crossplane/blob/master/cluster/charts/crossplane/templates/clusterrole.yaml

### CODE REVIEW

**Comment by negz** [comment]
I've tested this by:

1. Creating a `KubernetesCluster` that uses a `GKEClusterClass`.
1. Verifying that edits to the `GKECluster` connection secret are immediately propagated to the `KubernetesCluster` connection secret.
1. Creating a backup of all `KubernetesCluster`, `GKECluster`, and `Secret` resources using Velero.
1. Stopping provider-gcp
1. Manually deleting the `KubernetesCluster` and `GKECluster` by removing their finalizers
1. Starting provider-gcp
1. Restoring the backup made in step 3.
1. Verifying that edits to the `GKECluster` connection secret are immediately propagated to the `KubernetesCluster` connection secret.

File: pkg/meta/meta.go [hunk 1]
```diff
@@ -213,10 +226,69 @@ func WasCreated(o metav1.Object) bool {
...
+// The annotation name (which follows the prefix) can be anything that doesn't
+// collide with another annotation. to.propagation.crossplane.io/example would
+// be valid. This function uses a hash of the supplied object's namespace and
+// name in order to avoid collisions and keep the suffix relatively short.
+func AnnotationKeyPropagateTo(o metav1.Object) string {
+	// Writing to a hash never returns an error.
+	h := fnv.New32a()
+	h.Write([]byte(o.GetNamespace())) // nolint:errcheck
+	h.Write([]byte(o.GetName()))      // nolint:errcheck
+	return fmt.Sprintf("%s%x", AnnotationKeyPropagateToPrefix, h.Sum32())
```
Comment (hasheddan) [inline pkg/meta/meta.go:261]: Nice 👍

File: pkg/meta/meta.go [hunk 2]
```diff
@@ -32,10 +36,19 @@ import (
...
+// consents to propagation.
 const (
-	// ExternalNameAnnotationKey is the key in the annotations map of a resource
-	// for the name of the resource as it appears on provider's systems.
-	ExternalNameAnnotationKey = "crossplane.io/external-name"
+	AnnotationKeyPropagateToPrefix = "to.propagate.crossplane.io/"
+
+	AnnotationKeyPropagateFromNamespace = "from.propagate.crossplane.io/namespace"
+	AnnotationKeyPropagateFromName      = "from.propagate.crossplane.io/name"
+	AnnotationKeyPropagateFromUID       = "from.propagate.crossplane.io/uid"
```
Comment (hasheddan) [inline pkg/meta/meta.go:51]: Nit: is this used anywhere?

Comment (negz) [reply pkg/meta/meta.go:51]: It's not - good catch.

**Pull request merged**